hbase-client/src/main/java/org/apache/hadoop/hbase/CheckAndMutate.java

```diff
@@ -0,0 +1,312 @@
+    private final byte[] family;
+    private byte[] qualifier;
+    private CompareOperator op;
+    private byte[] value;
+    private TimeRange timeRange;
+
+    private Builder(byte[] row, byte[] family) {
+      this.row = Preconditions.checkNotNull(row, "row is null");
+      this.family = Preconditions.checkNotNull(family, "family is null");
+    }
+
+    /**
+     * @param qualifier column qualifier to check
+     * @return the builder object
+     */
+    public Builder qualifier(byte[] qualifier) {
+      this.qualifier = Preconditions.checkNotNull(qualifier, "qualifier is null. Consider using" +
+        " an empty byte array, or just do not call this method if you want a null qualifier");
+      return this;
+    }
+
+    /**
+     * @param timeRange timeRange to check
+     * @return the builder object
+     */
+    public Builder timeRange(TimeRange timeRange) {
+      this.timeRange = timeRange;
+      return this;
+    }
+
+    /**
+     * Check for lack of column
+     * @return the builder object
+     */
+    public Builder ifNotExists() {
+      this.op = CompareOperator.EQUAL;
+      this.value = null;
+      return this;
+    }
+
+    /**
+     * Check for equality
+     * @param value the expected value
+     * @return the builder object
+     */
+    public Builder ifEquals(byte[] value) {
+      return ifMatches(CompareOperator.EQUAL, value);
+    }
+
+    /**
+     * @param compareOp comparison operator to use
+     * @param value the expected value
+     * @return the builder object
+     */
+    public Builder ifMatches(CompareOperator compareOp, byte[] value) {
+      this.op = Preconditions.checkNotNull(compareOp, "compareOp is null");
+      this.value = Preconditions.checkNotNull(value, "value is null");
+      return this;
+    }
+
+    private void preCheck() {
+      Preconditions.checkNotNull(op, "condition is null. You need to specify the condition by" +
+        " calling ifNotExists/ifEquals/ifMatches before executing the request");
+    }
+
+    /**
+     * @param put data to put if check succeeds
+     * @return a CheckAndMutate object
+     */
+    public CheckAndMutate build(Put put) {
+      preCheck();
+      return new CheckAndMutate(row, family, qualifier, op, value, timeRange, put);
+    }
+
+    /**
+     * @param delete data to delete if check succeeds
+     * @return a CheckAndMutate object
+     */
+    public CheckAndMutate build(Delete delete) {
+      preCheck();
+      return new CheckAndMutate(row, family, qualifier, op, value, timeRange, delete);
+    }
+
+    /**
+     * @param mutation mutations to perform if check succeeds
+     * @return a CheckAndMutate object
+     */
+    public CheckAndMutate build(RowMutations mutation) {
+      preCheck();
+      return new CheckAndMutate(row, family, qualifier, op, value, timeRange, mutation);
+    }
+  }
+
+  /**
+   * A builder class for building a CheckAndMutate object with a filter.
+   */
+  public static class WithFilterBuilder {
```

Comment:
Interesting. Why you have to have a separate builder? I'm sure there is a reason for why can't be in above Builder.

Reply:
Would it help if this subclassed the above?

Reply:
Ah, I tried that, but looks like it's hard to subclass.